### PR TITLE
Bugfix: zero-initialized metrics for failed/aborted/skipped requests throwing off histograms

### PR DESCRIPTION
**Pull request by wseaton**
This fixes an issue reported by @kfirwolfson where the new finish reasons (and old ones like `ABORT`) cause incorrect stats to be recorded in histogram counters due to zero initialized values. 

Now we use a type state pattern to conditionally update them based on how far the request actually progressed through the pipeline.

### CODE REVIEW

File: vllm/v1/metrics/stats.py [hunk 1]
```diff
@@ -338,39 +367,44 @@ def update_from_finished_request(
...
+                queued_time=queued_time,
+                prefill_time=prefill_time,
+                decode_time=decode_time,
+                inference_time=inference_time,
+                mean_time_per_output_token=mean_tpot,
+            )
+        elif was_queued and was_scheduled:
+            # scheduled but no tokens (abort during prefill, KV error, etc.)
+            queued_time = req_stats.scheduled_ts - req_stats.queued_ts
+            timing = ScheduledTiming(queued_time=queued_time)
```
Comment (wseaton) [inline vllm/v1/metrics/stats.py]: The question here is does recording this way throw off any other reporting or muddy the waters semantically? Is it worth the extra cardinality of adding a label for finish reason?

File: vllm/v1/metrics/stats.py
Comment (kfirwolfson) [inline vllm/v1/metrics/stats.py:378]: I think this might also come into play and add incorrect stats of `first_token_latency`
I believe `is_prefilling` can be `True`, for example for requests which failed to load KV-Cache. Consider changing the above `if` statement to 
```python
if is_prefilling and num_new_generation_tokens > 0:
    #...
```
and adding a comment above it that explains why we test both

Comment (wseaton) [reply vllm/v1/metrics/stats.py:378]: pushed a fix and regression tests for this!

File: vllm/v1/metrics/stats.py [hunk 2]
```diff
@@ -411,39 +440,57 @@ def update_from_finished_request(
...
-        # Do not count the token generated by the prefill phase
-        mean_time_per_output_token = (
-            decode_time / (req_stats.num_generation_tokens - 1)
-            if req_stats.num_generation_tokens - 1 > 0
-            else 0
-        )
+        # build timing based on how far the request progressed
+        timing: RequestTiming = None
+        was_queued = req_stats.queued_ts > 0
+        was_scheduled = req_stats.scheduled_ts > 0
```
Comment (markmc) [inline vllm/v1/metrics/stats.py:452]: Honestly, I'm pretty sceptical of all of this new logic to determine "how far the request progressed"

The fix, in the end, is nothing more than:

```
if req_stats.queued_ts > 0 and req_stats.scheduled_ts > 0:
    queued_time = req_stats.scheduled_ts - req_stats.queued_ts
...
if finished_request.queued_time > 0:
    self.histogram_queue_time_request[engine_idx].observe(finished_request.queued_time)
```

AFAICT, we don't actually use this "how far the request progressed" information ...

File: vllm/v1/metrics/stats.py [hunk 3]
```diff
@@ -213,20 +213,49 @@ class RequestStateStats:
...
+    """Stats associated with a finished request.
+
+    The timing field uses its type to encode how far the
+    request progressed before finishing:
+    - None: rejected before scheduling (e.g., cache_threshold)
+    - ScheduledTiming: scheduled but no tokens generated (abort/error during prefill)
+    - CompletedTiming: generated at least one token (normal completion)
+    """
 
     finish_reason: "FinishReason"
```
Comment (markmc) [inline vllm/v1/metrics/stats.py:253]: I particularly don't like the way this data structure has this type-like field (`finish_reason`)

```
class FinishReason(enum.IntEnum):
    STOP = 0
    LENGTH = 1
    ABORT = 2
    ERROR = 3
```

which you might expect would determine which fields below are valid, but then we have this other `timing = None|scheduled|completed` thing which also indicates why/when the request was finished

We could make it all more coherent, but it doesn't seem like a bit deal to just say some of the fields may not be valid, and we'll only observe the fields that are valid

Some docs here explaining that some of the fields may not be valid would be nice though

File: vllm/v1/metrics/stats.py
Comment (markmc) [inline vllm/v1/metrics/stats.py:368]: This also assumes the prefill has completed ... ?

Comment (markmc) [inline vllm/v1/metrics/stats.py:364]: I'd restructure this method so that we process events and maybe `num_nans_in_logits`, then just return if `num_new_generation_tokens == 0`

Comment (markmc) [reply vllm/v1/metrics/stats.py:378]: You can see in `OutputProcessor.process_outputs()` that we assume the first `EngineCoreOutput` we receive for a request means that prefilling has completed

```
        for engine_core_output in engine_core_outputs:
            req_id = engine_core_output.request_id
            req_state = self.request_states.get(req_id)
	       if req_state is None:
                # Ignore output for already-aborted request.                                                                                                                                                                   
                continue

            # 1) Compute stats for this iteration.                                                                                                                                                                             
      	    self._update_stats_from_output(
                req_state, engine_core_output, engine_core_timestamp, iteration_stats
            )

            new_token_ids = engine_core_output.new_token_ids
            ...
            req_state.is_prefilling = False
```

I guess since #26813 that is obviously not a safe assumption anymore, but maybe also #9880?

In any case, I think we should start by addressing this assumption in `OutputProcessor`. And that feels like a separate PR

**Review by markmc — CHANGES_REQUESTED** [review]
Nice finds!
